### DISCUSSION TITLE
How to keep gettion location when application is in background?

### DISCUSSION THREAD

**Discussion by ComptonAlvaro** [dotnet/maui#15190] [Q&A]
I would like to have an application that get the location periodically, every 5 minutes for example, compare this location with some locations that I have stored (points of interest) and if the distance is less than x meters, then show a notification.

In the android documentation: https://developer.android.com/guide/components/foreground-services
 I have read that the foreground it is an option for applications likea music player. The music player is not running in the foreground most of the time, and is in the notification bar where you can see the song that is playing and another information.

Because of that, I would try to use a foreground service to see if it is a good option for me. This is the application:

[MauiForegroundServiceWithGps.zip](https://github.com/dotnet/maui/files/11519505/MauiForegroundServiceWithGps.zip)

the main code is in the AlarmService.cs file, in the method that do the work:

```
private async Task DoWorkAsync()
    {
        try
        {
            while (await _timer.WaitForNextTickAsync(_cts.Token))
            {
                GeolocationRequest miSolicitud = new GeolocationRequest(GeolocationAccuracy.Best);
                Location? myLocation = await Geolocation.Default.GetLocationAsync(miSolicitud);


                if (myLocation != null)
                {
                    double myDistance = Location.CalculateDistance(myLocation, _referenceLocation, DistanceUnits.Kilometers);
                    if (myDistance <= 1000)
                    {
                        _messenger.Send<TextMessage>(new TextMessage($"You are you {Math.Round(myDistance, 2)} kilometers from reference point."));
                    }
                }
            }
        }
        catch (System.OperationCanceledException ex)
        {
        }
    }
```
The problem with this is that when the application is in the background, this method stops. However, if I change the code of this method, intestad of request the location, for someting like this, it works if I put the application in the backgroud or if I close the application.

```
private int _notificationNumber = 0;
private async Task DoWorkAsync()
    {
        try
        {
            while (await _timer.WaitForNextTickAsync(_cts.Token))
            {
                _notificationNumber ++;
                _messenger.Send<TextMessage>(new TextMessage($"Number of notification: {_notificationNumber}"));
            }
        }
        catch (System.OperationCanceledException ex)
        {
        }
    }
```

So in general, the background can work as I expect, run although the application is in background or closed, but when I request the location it seems the bahaviour is different.

So my question is, how could I keep requesting the location although the application is in background or closed?

Thanks.

**Comment by Xyncgas** [comment]
If the platform doesn't let you what are you going to do, for example if Apple refuse letting applications get location in background because we are at war with spyware or china

**Comment by ComptonAlvaro** [comment]
In my case the application is an Android application that runs in Android 12.

**Comment by janseris** [comment]
Try this, I think it was the thing that helped me (if you use Android): https://github.com/jfversluis/XFBackgroundLocationSample
You need to declare it deep in Android platform code - a _foreground service_ (= background service).

**Comment by ComptonAlvaro** [comment]
@janseris Thanks for the example.

I have tried this code and it happens the same to me. If I put the application in the backgroud or I close it, the location of the GPS is not requested. If I open the application again, then it continues requesting the location.

If we focus in the code of the location service:

```
public async Task Run(CancellationToken token)
    {
        await Task.Run(async () => {
            while (!stopping)
            {
                token.ThrowIfCancellationRequested();
                try
                {
                    await Task.Delay(2000);
                    var request = new GeolocationRequest(GeolocationAccuracy.High);
                    var location = await Geolocation.GetLocationAsync(request);
                    if (location != null)
                    {
                        var message = new LocationMessage
                        {
                            Latitude = location.Latitude,
                            Longitude = location.Longitude
                        };
                        Device.BeginInvokeOnMainThread(() =>
                        {
                            MessagingCenter.Send(message, "Location");
                        });
                    }
                }
                catch (Exception ex)
                {
                    Device.BeginInvokeOnMainThread(() =>
                    {
                        var errormessage = new LocationErrorMessage();
                        MessagingCenter.Send(errormessage, "LocationError");
                    });
                }
            }
            return;
        }, token);
    }
```

I have realized that if I set a breakpint in the delay of 2 seconds, if I set the application in background or I stop it, the code is not run, although i still can see the icon in the state bar that the foreground services is running.

I have tried to don't give to the application permission to can use the location. In this case, in this code, the exception is thrown. I have realized that in this case, if I put the application in the background or I close the application, the code is still running. I know it because of the breakpint in the delay of 2 seconds.

So it works in the same way than in my code. If the application has permissions to use the location, the code is not run when the application is not in the foreground, although the foreground services is still running. However, if the location is no request, for example because an expcetion is thrown, the the code is running always, no matter if the application is in the background or not.

**Reply by janseris** [reply]
Have you done the Android Platform code with some notificatio and Intent and something? You cannot influence the functionality from the code you sent itself. You should get a notification on Android when you start the "foreground service" (=background service) which will show "application xxx will now run in background" or something. If I remember browsing my code correctly, you can only start the service with firing the notification at the same time.

**Comment by ComptonAlvaro** [comment]
Yes, the first notification that shows that the foreground is running it is shown. It is not possible to create and start a fororeground without a notification. I know that this first notification is created it is because in the status bar I can see the icon of my application and also I can see the notification in the notifications panel.

I know that the location is requested because every time that the location is requested, an arrow icon is shown in the icons beside the time, in the upper right corner. So this icon is show every 2 seconds (the interval of the retrieving). How ever, this icon is not shown when I put the application in the backgroud or I close it.

Also I know that the location is not request when the application is not open because if i set a breakpoint in code that request the location, this breakpoint is not reached.

**Reply by HaimoHeymann** [reply]
Did you get a working solution? I also struggle for a long time with reading the geolocation when my app is in background. 
All solutions I found **stop as soon as my app is in background**

**Reply by janseris** [reply]
@HaimoHeymann see my posts above (Android only)

**Comment by ComptonAlvaro** [comment]
@HaimoHeymann I have created a repository with the code: [repository](https://github.com/ComptonAlvaro/MauiGpsRequestInForeground).

The repository is just the code of classes that make it possible, but it is not complete, because it is from a project that I can't share the whole code. Anyway, it shows the part of code that you need.

As important points, the code shows how to use a foreground that use a timer inside a while to request the location periodically. It is in the class "ServicioAlarmas.cs".

Also the codes shows how to start the foreground when the device is rebooted. For that, you could check the code in the \platforms\Android directory.

There are another helpers, for example, the helper that request the location. I hope this classes work.

Any question that you could have, tell me and I will try to help you.

I hope this code can help you.

**Reply by HaimoHeymann** [reply]
Great! It will help to get the GPS Listener run

**Reply by ComptonAlvaro** [reply]
I am glad it helps you and the code that I share is enough.

Thanks for tell it could help you, it is nice to can help if I can.
